src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs

```diff
@@ -817,44 +840,88 @@ public void ReceivingMessage()
+            {
+                var encoderFactory = new BinaryMessageEncoderFactory(stream, _messageEncoderSettings);
+                ResponseMessage responseMessage;
+                if (stream.IsCompressedMessage())
+                {
+                    var selector = new CompressedMessageEncoderSelector(_compressorFactory);
+                    var compressedMessage = (CompressedMessage)selector.GetEncoder(encoderFactory).ReadMessage();
+                    responseMessage = (ResponseMessage)compressedMessage.OriginalMessage;
+                }
+                else
+                {
+                    responseMessage = (ResponseMessage)encoderSelector.GetEncoder(encoderFactory).ReadMessage();
+                }
+
+                return responseMessage;
+            }
         }
 
         private class SendMessagesHelper
         {
+            private static readonly HashSet<string> __messagesNotToBeCompressed = new HashSet<string>
+            {
+                "isMaster",
+                "saslStart",
+                "saslContinue",
+                "getnonce",
+                "authenticate",
+                "createUser",
+                "updateUser",
+                "copydbsaslstart",
+                "copydbgetnonce",
+                "copydb"
+            };
+
+            private readonly ICompressorFactory _compressorFactory;
             private readonly Stopwatch _commandStopwatch;
             private readonly BinaryConnection _connection;
             private readonly MessageEncoderSettings _messageEncoderSettings;
             private readonly List<RequestMessage> _messages;
             private Lazy<List<int>> _requestIds;
             private TimeSpan _serializationDuration;
             private Stopwatch _networkStopwatch;
-
-            public SendMessagesHelper(BinaryConnection connection, IEnumerable<RequestMessage> messages, MessageEncoderSettings messageEncoderSettings)
+            private readonly CompressorType? _sendCompressorType;
+
+            public SendMessagesHelper(
+                BinaryConnection connection,
+                IEnumerable<RequestMessage> messages,
+                CompressorType? sendCompressorType,
+                ICompressorFactory compressorFactory,
+                MessageEncoderSettings messageEncoderSettings)
             {
+                _compressorFactory = compressorFactory;
                 _connection = connection;
                 _messages = messages.ToList();
                 _messageEncoderSettings = messageEncoderSettings;
 
                 _commandStopwatch = Stopwatch.StartNew();
                 _requestIds = new Lazy<List<int>>(() => _messages.Select(m => m.RequestId).ToList());
+                _sendCompressorType = sendCompressorType;
             }
 
-            public IByteBuffer EncodeMessages(CancellationToken cancellationToken)
+            public Tuple<IByteBuffer, IByteBuffer> EncodeMessages(CancellationToken cancellationToken)
```

Comment:
NOTE: this method can be rewritten in more good looking way, if we add the Tuple nuget to `core` library: System.ValueTuple. The code will look like: 

    public (IByteBuffer efficientMessageBuffer, IByteBuffer originalMessageBuffer) EncodeMessages(CancellationToken cancellationToken)

and using will be:

    var buffers = helper.EncodeMessages(cancellationToken);
    using (var efficientMessageBuffer = buffers.efficientMessageBuffer)
    using (var originalMessageBuffer = buffers.originalMessageBuffer)

Reply:
I don't like the oldest way too.
But the new version looks really not bad and helps to write code simpler/smaller and as a result clearer, at least for small cases like this. 
Also 4-5 new code lines will unnecessary distract attention.
But if you really disagree, I don't mind.

Reply:
> Also 4-5 new code lines will unnecessary distract attention.

I disagree with that statement. A small value class to hold multiple return values is not a distraction. It's better than using Item1 and Item2!

Reply:
I don't want to depend on an external package for Tuple, even if it is a Microsoft package.

Just make a small class to hold your compound return value. It's 3 lines of code plus braces.